source/agora/test/Base.d

```diff
@@ -36,6 +39,70 @@ import std.algorithm.iteration;
 import std.exception;
 import std.format;
 
+/*******************************************************************************
+
+    Create a set of chained transactions, where each transaction spends the
+    entire sum of the previous transaction.
+
+    Useful for quickly generating transactions for use with unittests.
+
+    Params:
+        root = the first transaction in the chain will spend the output of
+               this transaction
+        count = the number of transactions to create
+        key_pair = the key pair used to sign transactions and to send
+                   the output to
+
+*******************************************************************************/
+
+public Transaction[] getChainedTransactions (Transaction root, size_t count,
+    KeyPair key_pair)
+{
+    Transaction[] transactions;
+    Hash last_tx_hash = hashFull(root);
+
+    foreach (idx; 0 .. count)
+    {
+        Transaction tx =
+        {
+            [Input(last_tx_hash, 0)],
+            [Output(40_000_000, key_pair.address)]  // send to the same address
+        };
+
+        auto signature = key_pair.secret.sign(hashFull(tx)[]);
+        tx.inputs[0].signature = signature;
+        last_tx_hash = hashFull(tx);
+        transactions ~= tx;
+    }
+
+    return transactions;
+}
+
+/*******************************************************************************
+
+    Create a set of chained blocks, containing the provided transactions,
+    and linking back to the provided root block.
+    Note that the root block is not contained in the array,
+    as it's already available at the call site.
+
+    Params:
+        root = the first block in the chain
+        txes = the list of transactions to create the blocks for
+
+*******************************************************************************/
+
+public Block[] getChainedBlocks (Block root, Transaction[] txes)
+{
+    Block[] blocks;
+
+    foreach (tx; txes)
+    {
+        root = root.makeNewBlock(tx);
+        blocks ~= root;
```

Comment:
I have a feeling you deleted some code here, because the `root` assignment is useless

Reply:
Perhaps it's UFCS that's confusing you.

```D
root = makeNewBlock(root, tx);
```

Reply:
I will make it clearer maybe by adding a `temp` block up there?

Reply:
`blocks ~= blocks[$-1].makeNewBlock(tx)` ? Ofc you need to handle the first loop, but...

Reply:
`temp` is also good tho